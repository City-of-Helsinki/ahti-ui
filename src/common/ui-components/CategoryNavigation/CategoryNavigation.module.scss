@import '../../../styles/icons';

.categoryContainer {
  display: flex;
  flex-direction: column;
  align-items: center;
  cursor: pointer;
}

.categoryIconContainer {
  display: flex;
  align-items: center;
  justify-content: center;
  background-color: var(--hds-brand-color-fog);
  width: 50px;
  height: 50px;
  border-radius: 50px;
}

.categoryIconSelected {
  background-color: var(--hds-brand-color-fog-dark-50);
}

.categoryName {
  margin-top: 5px;
  font-size: 10px;
  line-height: 13px;
  font-weight: 500;
}
