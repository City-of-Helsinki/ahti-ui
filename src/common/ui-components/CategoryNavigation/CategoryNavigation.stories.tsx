import React from 'react';

import CategoryNavigation from './CategoryNavigation';

export default {
  title: 'CategoryNavigation',
  component: CategoryNavigation,
};

const noop = () => {
  /* NOOP */
};

export const MultipleCategories = () => (
  <CategoryNavigation
    categories={[
      { id: 'ahti:category:island', name: 'Islands' },
      { id: 'ahti:category:island', name: 'Islands' },
      { id: 'ahti:category:island', name: 'Islands' },
      { id: 'ahti:category:island', name: 'Islands' },
      { id: 'ahti:category:island', name: 'Islands' },
      { id: 'ahti:category:island', name: 'Islands' },
      { id: 'ahti:category:island', name: 'Islands' },
    ]}
    onClick={noop}
  />
);

MultipleCategories.story = {
  name: 'Multiple categories',
};

export const SingleCategory = () => (
  <CategoryNavigation
    categories={[{ id: 'ahti:category:island', name: 'Islands' }]}
    onClick={noop}
  />
);

SingleCategory.story = {
  name: 'Single category',
};

export const MaxItemsCategories = () => (
  <CategoryNavigation
    categories={[
      { id: 'ahti:category:island', name: 'Islands' },
      { id: 'ahti:category:island', name: 'Islands' },
      { id: 'ahti:category:island', name: 'Islands' },
    ]}
    maxDisplayedCategories={3}
    onClick={noop}
  />
);

MaxItemsCategories.story = {
  name: 'Max items # of categories',
};
