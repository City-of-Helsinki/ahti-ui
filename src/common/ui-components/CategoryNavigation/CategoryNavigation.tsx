import React from 'react';
import Slider from 'react-slick';
import classNames from 'classnames/bind';
import { useTranslation } from 'react-i18next';

import CategoryIcon from '../CategoryIcon/CategoryIcon';
import { Filter } from '../../../../alltypes';
import styles from './CategoryNavigation.module.scss';

import './CategoryNavigation.scss';

const cx = classNames.bind(styles);

export interface CategoryNavigationProps {
  readonly className?: string;
  readonly categories: Filter[];
  readonly selectedId?: string;
  readonly maxDisplayedCategories?: number;
  readonly translated?: boolean;
  onClick(categoryId: string): void;
}

const CategoryNavigation: React.FC<CategoryNavigationProps> = ({
  className,
  categories,
  selectedId,
  maxDisplayedCategories = 5,
  translated = false,
  onClick,
}) => {
  const { t, i18n } = useTranslation();
  const sliderSettings = {
    speed: 500,
    slidesToShow:
      categories.length > maxDisplayedCategories
        ? maxDisplayedCategories
        : categories.length,
    slidesToScroll: 1,
  };

  const enterPressed = (event: React.KeyboardEvent): boolean => {
    const code = event.keyCode || event.which;
    return code === 13;
  };

  return (
    <Slider
      className={cx('categoryNavigationSlider', className)}
      {...sliderSettings}
    >
      {categories.map((category, id: number) => (
        <div key={id}>
          <div
            role={'button'}
            className={cx(styles.categoryContainer, {
              categoryIconSelected: category.id === selectedId,
            })}
            onClick={() => onClick(category.id)}
            onKeyPress={(event) => enterPressed(event) && onClick(category.id)}
            tabIndex={0}
          >
            <div className={styles.categoryIconContainer}>
              <CategoryIcon category={category.id} className={styles.bigIcon} />
            </div>
            <div className={styles.categoryName}>
              {translated && i18n.exists(`categories_and_tags.${category.id}`)
                ? t(`categories_and_tags.${category.id}`)
                : category.name}
            </div>
          </div>
        </div>
      ))}
    </Slider>
  );
};

export default CategoryNavigation;
