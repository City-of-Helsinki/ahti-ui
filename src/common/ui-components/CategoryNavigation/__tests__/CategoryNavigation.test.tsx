import React from 'react';
import { mount, shallow } from 'enzyme';

import CategoryNavigation from '../CategoryNavigation';

describe('CategoryNavigation', () => {
  const mockCategory = {
    id: 'ahti:category:island',
    name: 'Islands'
  };

  it('renders without crashing', () => {
    const wrapper = shallow(
      <CategoryNavigation categories={[mockCategory]} onClick={jest.fn} />
    );
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders slick-next when number of categories > maxDisplayedCategories"', () => {
    const wrapper = mount(
      <CategoryNavigation
        categories={Array(10).fill(mockCategory)}
        maxDisplayedCategories={5}
        onClick={jest.fn}
      />
    );
    expect(wrapper.find('.slick-next').exists()).toBeTruthy();
  });

  it("doesn't render slick-next when number of categories <= maxDisplayedCategories", () => {
    const wrapper = mount(
      <CategoryNavigation
        categories={Array(3).fill(mockCategory)}
        maxDisplayedCategories={5}
        onClick={jest.fn}
      />
    );
    expect(wrapper.find('.slick-next').exists()).toBeFalsy();
  });

  it('calls onClick with id of clicked category', () => {
    const mockOnClick = jest.fn();
    const wrapper = shallow(
      <CategoryNavigation categories={[mockCategory]} onClick={mockOnClick} />
    );
    wrapper.find('.categoryContainer').simulate('click');
    expect(mockOnClick).toBeCalledWith(mockCategory.id);
  });
});
