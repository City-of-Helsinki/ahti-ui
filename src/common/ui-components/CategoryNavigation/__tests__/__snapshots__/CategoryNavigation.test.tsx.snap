// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoryNavigation renders without crashing 1`] = `"<div class=\\"slick-slider categoryNavigationSlider slick-initialized\\"><div class=\\"slick-list\\"><div class=\\"slick-track\\" style=\\"width:100%;left:0%\\"><div data-index=\\"0\\" class=\\"slick-slide slick-active slick-current\\" tabindex=\\"-1\\" aria-hidden=\\"false\\" style=\\"outline:none;width:100%\\"><div><div tabindex=\\"-1\\" style=\\"width:100%;display:inline-block\\"><div role=\\"button\\" class=\\"categoryContainer\\" tabindex=\\"0\\"><div class=\\"categoryIconContainer\\"><svg class=\\"bigIcon\\">icon-location.svg</svg></div><div class=\\"categoryName\\">Islands</div></div></div></div></div></div></div></div>"`;
